File: app/src/main/java/ai/elimu/analytics/receiver/LetterSoundLearningEventReceiver.java
package ai.elimu.analytics.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.provider.Settings;

import java.util.Calendar;

import ai.elimu.analytics.dao.LetterSoundLearningEventDao;
import ai.elimu.analytics.db.RoomDb;
import ai.elimu.analytics.entity.LetterSoundCorrespondenceLearningEvent;
import timber.log.Timber;

public class LetterSoundLearningEventReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        Timber.i("onReceive");

        String androidId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
        Timber.i("androidId: \"" + androidId + "\"");

        String packageName = intent.getStringExtra("packageName");
        Timber.i("packageName: \"" + packageName + "\"");

        Calendar timestamp = Calendar.getInstance();
        Timber.i("timestamp.getTime(): " + timestamp.getTime());

        Long letterSoundId = null;
        if (intent.hasExtra("letterSoundId")) {
            letterSoundId = intent.getLongExtra("letterSoundId", 0);
        }
        Timber.i("letterSoundId: " + letterSoundId);

        String[] letterSoundLetterTexts = null;
        if (intent.hasExtra("letterSoundLetterTexts")) {
            letterSoundLetterTexts = intent.getStringArrayExtra("letterSoundLetterTexts");
        }
        Timber.i("letterSoundLetterTexts: " + letterSoundLetterTexts);

        String[] letterSoundSoundValuesIpa = null;
        if (intent.hasExtra("letterSoundSoundValuesIpa")) {
            letterSoundSoundValuesIpa = intent.getStringArrayExtra("letterSoundSoundValuesIpa");
        }
        Timber.i("letterSoundSoundValuesIpa: " + letterSoundSoundValuesIpa);

        LetterSoundCorrespondenceLearningEvent letterSoundLearningEvent = new LetterSoundCorrespondenceLearningEvent();
        letterSoundLearningEvent.setAndroidId(androidId);
        letterSoundLearningEvent.setPackageName(packageName);
        letterSoundLearningEvent.setTime(timestamp);
        letterSoundLearningEvent.setLetterSoundCorrespondenceId(letterSoundId);
        // TODO: set letterSoundLetterTexts
        // TODO: set letterSoundSoundValuesIpa

        // Store in database
        RoomDb roomDb = RoomDb.getDatabase(context);
        LetterSoundLearningEventDao letterSoundLearningEventDao = roomDb.letterSoundLearningEventDao();
        RoomDb.databaseWriteExecutor.execute(() -> {
            letterSoundLearningEventDao.insert(letterSoundLearningEvent);
        });
    }
}

File: app/src/main/java/ai/elimu/analytics/receiver/LetterSoundLearningEventReceiver.kt
package ai.elimu.analytics.receiver

import ai.elimu.analytics.db.RoomDb
import ai.elimu.analytics.entity.LetterSoundLearningEvent
import ai.elimu.analytics.utils.BundleKeys
import ai.elimu.analytics.utils.research.ExperimentAssignmentHelper
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import timber.log.Timber
import java.util.Calendar

class LetterSoundLearningEventReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        Timber.i("onReceive")

        val bundle = intent.extras
        bundle?.let {
            for (key in bundle.keySet()) {
                val value = bundle.get(key)
                Timber.i("${key}=${value}")
            }
        }

        try {
            val event = LetterSoundLearningEvent()

            val androidId: String = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
            event.androidId = androidId

            val packageName: String = intent.getStringExtra(BundleKeys.KEY_PACKAGE_NAME) ?: ""
            if (TextUtils.isEmpty(packageName)) {
                throw IllegalArgumentException("Missing packageName")
            }
            event.packageName = packageName

            val timestamp: Calendar = Calendar.getInstance()
            event.timestamp = timestamp

            val researchExperiment = ExperimentAssignmentHelper.CURRENT_EXPERIMENT
            event.researchExperiment = researchExperiment

            val experimentGroup = ExperimentAssignmentHelper.getExperimentGroup(context)
            event.experimentGroup = experimentGroup

            val additionalData: String? = intent.getStringExtra(BundleKeys.KEY_ADDITIONAL_DATA)
            event.additionalData = additionalData

            val letterSoundLetters: ArrayList<String> = intent.getStringArrayListExtra(BundleKeys.KEY_LETTER_SOUND_LETTERS) ?: arrayListOf()
            if (letterSoundLetters.isEmpty()) {
                throw IllegalArgumentException("Missing letterSoundLetters")
            }
            event.letterSoundLetters = letterSoundLetters

            val letterSoundSounds: ArrayList<String> = intent.getStringArrayListExtra(BundleKeys.KEY_LETTER_SOUND_SOUNDS) ?: arrayListOf()
            if (letterSoundSounds.isEmpty()) {
                throw IllegalArgumentException("Missing letterSoundSounds")
            }
            event.letterSoundSounds = letterSoundSounds

            val letterSoundId: Long = intent.getLongExtra(BundleKeys.KEY_LETTER_SOUND_ID, 0)
            if (letterSoundId > 0) {
                event.letterSoundId = letterSoundId
            }

            // Store the event in the database
            val roomDb = RoomDb.getDatabase(context)
            RoomDb.databaseWriteExecutor.execute {
                roomDb.letterSoundLearningEventDao().insert(event)
            }
        } catch (e: Exception) {
            Timber.e(e)
            val results: Bundle = getResultExtras(true)
            val errorMessage = e.message ?: e::class.simpleName
            results.putString("errorMessage", errorMessage)
        }
    }
}
